// stdf-parser/lib/records/tsr.record.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsrRecord = void 0;
const constants_1 = require("../constants");
const stdf_record_1 = require("../stdf-record");
class TsrRecord extends stdf_record_1.StdfRecord {
    headNum;
    siteNum;
    testType;
    testNum;
    execCount;
    failCount;
    alarmCount;
    testName;
    seqName;
    testLabel;
    optFlag;
    testTime;
    testMin;
    testMax;
    testSums;
    testSquares;
    constructor(type, sub, br) {
        super(type, sub, br);
        this.headNum = br.readUInt8();
        this.siteNum = br.readUInt8();
        this.testType = br.readChar();
        this.testNum = br.readUInt32();
        let count;
        this.execCount = (count = br.readUInt32()) === constants_1.MAX_UINT32 ? undefined : count;
        this.failCount = (count = br.readUInt32()) === constants_1.MAX_UINT32 ? undefined : count;
        this.alarmCount = (count = br.readUInt32()) === constants_1.MAX_UINT32 ? undefined : count;
        this.testName = br.readVarChar();
        this.seqName = br.readVarChar();
        this.testLabel = br.readVarChar();
        this.optFlag = br.readUInt8();
        this.testTime = br.readFloat();
        this.testMin = br.readFloat();
        this.testMax = br.readFloat();
        this.testSums = br.readFloat();
        this.testSquares = br.readFloat();
    }
}
exports.TsrRecord = TsrRecord;
