// stdf-parser/lib/records/wrr.record.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WrrRecord = void 0;
const stdf_record_1 = require("../stdf-record");
class WrrRecord extends stdf_record_1.StdfRecord {
    headNum;
    siteGrp;
    finishTime;
    partCount;
    retestCount;
    abortCount;
    goodCount;
    funcCount;
    waferId;
    fabWaferId;
    frameId;
    maskId;
    userDesc;
    execDesc;
    constructor(type, sub, br) {
        super(type, sub, br);
        this.headNum = br.readInt8();
        this.siteGrp = br.readInt8();
        this.finishTime = br.readDate();
        this.partCount = br.readUInt32();
        this.retestCount = br.readUInt32();
        this.abortCount = br.readUInt32();
        this.goodCount = br.readUInt32();
        this.funcCount = br.readUInt32();
        this.waferId = br.readVarChar();
        this.fabWaferId = br.readVarChar();
        this.frameId = br.readVarChar();
        this.maskId = br.readVarChar();
        this.userDesc = br.readVarChar();
        this.execDesc = br.readVarChar();
    }
}
exports.WrrRecord = WrrRecord;
